src/test/java/io/github/mapepire_ibmi/PoolTest.java

```diff
@@ -11,49 +15,273 @@
+
+        SqlJob job = poolSpy.getJob();
+        assertEquals(JobStatus.Ready, job.getStatus());
+        assertEquals(0, job.getRunningCount());
+        assertEquals(2, poolSpy.getActiveJobCount());
+        verify(poolSpy, times(0)).addJob();
+
+        future.join();
+        poolSpy.end();
+    }
+
+    @Test
+    void getJobGivesFreeJob() throws Exception {
+        PoolOptions options = new PoolOptions(MapepireTest.getCreds(), 2, 2);
+        Pool poolSpy = spy(new Pool(options));
+        poolSpy.init().get();
+        assertEquals(2, poolSpy.getActiveJobCount());
+
+        List<CompletableFuture<QueryResult<Object>>> futures = new ArrayList<>();
+        for (int i = 0; i < 3; i++) {
+            futures.add(poolSpy.execute("SELECT * FROM SAMPLE.SYSCOLUMNS"));
+        }
+        CompletableFuture<Void> allFutures = CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
+
+        SqlJob job = poolSpy.popJob().get();
+        assertEquals(JobStatus.Ready, job.getStatus());
+        assertEquals(1, job.getRunningCount());
+        assertEquals(2, poolSpy.getActiveJobCount());
+        verify(poolSpy, times(0)).addJob();
+
+        allFutures.join();
+        poolSpy.end();
+    }
+
+    @Test
+    void getJobAddsNewJob() throws Exception {
+        PoolOptions options = new PoolOptions(MapepireTest.getCreds(), 3, 2);
+        Pool poolSpy = spy(new Pool(options));
+        poolSpy.init().get();
+        assertEquals(2, poolSpy.getActiveJobCount());
+
+        List<CompletableFuture<QueryResult<Object>>> futures = new ArrayList<>();
+        for (int i = 0; i < 4; i++) {
+            futures.add(poolSpy.execute("SELECT * FROM SAMPLE.SYSCOLUMNS"));
+        }
+        CompletableFuture<Void> allFutures = CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
+
+        SqlJob job = poolSpy.getJob();
+        assertEquals(JobStatus.Ready, job.getStatus());
+        assertEquals(0, job.getRunningCount());
+        assertEquals(3, poolSpy.getActiveJobCount());
+        verify(poolSpy, times(1)).addJob();
+
+        allFutures.join();
+        poolSpy.end();
+    }
+
+    @Test
+    void getJobDoesNotExceedMaxSize() throws Exception {
+        PoolOptions options = new PoolOptions(MapepireTest.getCreds(), 2, 2);
+        Pool poolSpy = spy(new Pool(options));
+        poolSpy.init().get();
+        assertEquals(2, poolSpy.getActiveJobCount());
+
+        List<CompletableFuture<QueryResult<Object>>> futures = new ArrayList<>();
+        for (int i = 0; i < 4; i++) {
+            futures.add(poolSpy.execute("SELECT * FROM SAMPLE.SYSCOLUMNS"));
+        }
+        CompletableFuture<Void> allFutures = CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
+
+        SqlJob job = poolSpy.getJob();
+        assertEquals(JobStatus.Ready, job.getStatus());
+        assertEquals(2, job.getRunningCount());
+        assertEquals(2, poolSpy.getActiveJobCount());
+        verify(poolSpy, times(1)).addJob();
```

Comment:
Why are we expecting that addJob is called here? It appears that the pool is already full

Reply:
I will make these fixes in a separate PR once https://github.com/Mapepire-IBMi/mapepire-java/issues/43 is fixed